### ISSUE TITLE
RWLock is not really a RWLock?

### ISSUE THREAD

**Issue by margaretbloom** [PaloAltoNetworks/minemeld-core#295]
I don't know the rationale behind the `RWLock` class in `ft/util.py` but it doesn't behave as the standard Readers-Writers lock, particularly the sequence RL-WL-RL creates a deadlock.  
If a writer gets into two readers trying to acquire the lock the system deadlocks (*gevent* can detect this condition and throw an exception).

Steps to reproduce:

```
lock = RWLock()
lock.rlock()
gevent.spawn(lambda: lock.lock())
gevent.sleep(1)
lock.rlock()
```
If the idea was to create a standard RW lock and if this class is used/can be used in critical parts of the application it's better to fix it once and for all since race conditions are nasty to debug and anyway the RW lock is a textbook algorithm so little effort is required.  

I'm available for patching once the use cases for `RWLock` are explicated.  
I've backed off as I've not yet completed a full review of the code.

**Comment by jtschichold** [comment]
Hi @margaretbloom,
thanks for looking into this. `RWLock` is an implementation of _Write-preferring_ Readers-Writer lock, where additional Readers can't lock if the Writer is waiting. That is the reason of the deadlock in your sample:

1. main glet read locks
2. writer glet tries to write lock. This results in writer glet waiting on main glet
3. main glet tries to read lock again without freeing the first read lock. This results in main glet waiting on writer glet - deadlock

The second read lock (3) waits on writer glet because this is a Writer-preferring RW Lock. In a classic RW Lock the second read lock operation wouldn't wait for the writer.

If the second read lock is performed in a glet there is no deadlock:

```
from minemeld.ft.utils import RWLock
import gevent

def wglet():
    try:
        print 'Acquiring write lock'
        lock.lock()
        print 'Write lock acquired'
    except Exception as e:
        print 'Exception in wglet', e

def rglet():
    try:
        print 'Acquiring read lock'
        lock.rlock()
        print 'Read lock acquired'
    except Exception as e:
        print 'Exception in rglet', e

lock = RWLock()

lock.rlock()

wg = gevent.spawn(wglet)
gevent.sleep(0.1)
r2g = gevent.spawn(rglet)
gevent.sleep(0.1)

gevent.sleep(10.0)
```

Is this what you would expect from a Writer-preferring RW Lock?

Thanks

**Comment by margaretbloom** [comment]
Ah, OK, sorry my bad.
It's simply not reentrant but that's not the usual requirement for a Lock anyway.
Slightly off topic: what's the purpose of the `m3` mutex? wouldn't `r` suffice?

**Comment by jtschichold** [comment]
You are right, `r` should be enough. `m3` is dead code from a previous test.

**Comment by jtschichold** [comment]
please, let me know if you want to send in a PR
